libs/domain/cart/coupon/coupon.component.ts

```diff
@@ -0,0 +1,128 @@
+import { CartComponentMixin, CartService } from '@spryker-oryx/cart';
+import { resolve } from '@spryker-oryx/di';
+import { ContentMixin } from '@spryker-oryx/experience';
+import { LocaleService } from '@spryker-oryx/i18n';
+import { NotificationService } from '@spryker-oryx/site';
+import { AlertType } from '@spryker-oryx/ui';
+import { ButtonSize, ButtonType } from '@spryker-oryx/ui/button';
+import { IconTypes } from '@spryker-oryx/ui/icon';
+import { ColorType } from '@spryker-oryx/ui/link';
+import { ClearIconAppearance } from '@spryker-oryx/ui/searchbox';
+import { hydrate } from '@spryker-oryx/utilities';
+import { LitElement, TemplateResult, html } from 'lit';
+import { query, state } from 'lit/decorators.js';
+import { repeat } from 'lit/directives/repeat.js';
+import { tap } from 'rxjs';
+import { couponStyles } from './coupon.styles';
+
+@hydrate({ event: ['window:load'] })
+export class CouponComponent extends CartComponentMixin(
+  ContentMixin(LitElement)
+) {
+  static styles = couponStyles;
+
+  @state() hasError = false;
+  @state() errorMessageDescription: '' | string = '';
+
+  @query('input[name=coupon]') coupon?: HTMLInputElement;
+
+  protected cartService = resolve(CartService);
+  protected notificationService = resolve(NotificationService);
+  protected localeService = resolve(LocaleService);
+
+  protected override render(): TemplateResult | void {
+    if (this.$isEmpty()) {
+      return;
+    }
+
+    this.errorMessageDescription =
+      this.coupon?.value === ''
+        ? `${this.i18n('coupon.insert-a-coupon')}`
+        : `${this.i18n('coupon.cart-code-can-not-be-added')}`;
+
+    return html`
+      <oryx-layout>
+        <h3>${this.i18n('coupon.have-a-coupon')}?</h3>
+
+        <section>
+          <oryx-input
+            ?hasError="${this.hasError}"
+            .errorMessage="${this.hasError ? this.errorMessageDescription : ''}"
+          >
+            <input placeholder="Coupon code" name="coupon" />
+            <oryx-icon
+              type=${IconTypes.Close}
+              class="clear-icon"
+              appearance=${ClearIconAppearance.Hover}
+              @click=${(): void => this.clear()}
+            ></oryx-icon>
+          </oryx-input>
+          <oryx-button
+            .type="${ButtonType.Outline}"
+            .color="${ColorType.Neutral}"
+            .size=${ButtonSize.Md}
+            @click=${this.onSubmit}
+          >
+            ${this.i18n('coupon.apply')}
+          </oryx-button>
+        </section>
+      </oryx-layout>
+
+      <ul>
+        ${repeat(
+          this.$coupons(),
+          (coupon) => coupon.code,
+          (coupon) => {
+            let expirationDateTime = '';
+            this.localeService
+              .formatDate(coupon.expirationDateTime)
+              .pipe(
+                tap((date) => {
+                  expirationDateTime = date ?? '';
+                })
+              )
+              .subscribe();
```

Comment:
My suggestion is moving this logic into separate computed value in order not to use observables inside template.
ALSO, using combination of sync value + async subscription is BAD pattern, please avoid doing it

Reply:
Definitely, i use  this quickly on component to check design alignment and forget to refactor. thks